### PR TITLE
Update writeChunked method in bytestream.go to allow resumable upload with ByteStream Write API.

### PR DESCRIPTION
**Pull request by YukaSadaoka**
Update writeChunked method in bytestream.go to allow resumable upload with ByteStream Write API.

I also added a new struct ResumableWriteOpts and it’s added to Client to keep track of write offset so that we can start writing data at an arbitrary offset.

### CODE REVIEW

File: go/pkg/client/client.go [hunk 1]
```diff
@@ -474,6 +475,17 @@ func getRPCCreds(ctx context.Context, credFile string, useApplicationDefault boo
 	return rpcCreds, CredsFileAuth, nil
 }
 
+// ResumableWriteOpts sets options required for resumable ByteStream write operations.
+type ResumableWriteOpts struct {
+	FinishWrite       bool
+	LastLogicalOffset int64
```
Comment (ramymedhat) [inline go/pkg/client/client.go:481]: I'm not sure this is what you want to do here. This is an option that applies to the entire client, not just a single request. There can be many individual bytestream requests in flight for the same client.

Comment (YukaSadaoka) [reply go/pkg/client/client.go:480]: Thanks for the input. Updated the code. 
I removed the field 'ResumableWriteOpts' from Client struct in client.go. I wanted to make the code backward compatible, so I removed changes from the existing method(writeChunked) in bytestream.go and I added new methods taking write options such as offset and a boolean flag for determining to finish a write operation. Please let me know what you think.

File: go/pkg/client/bytestream.go [hunk 2]
```diff
@@ -73,6 +73,69 @@ func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chun
...
+		return 0, err
+	}
+	return c.writeChunkedWithOffset(ctx, name, ch, opts)
+}
+
+// writeChunkedWithOffset uploads chunked data with a given resource name to the CAS at an arbitrary offset.
+func (c *Client) writeChunkedWithOffset(ctx context.Context, name string, ch *chunker.Chunker, opts *ByteStreamWriteOpts) (int64, error) {
+	var totalBytes int64
+	var offset = opts.LastLogicalOffset
+	var initialOffset = opts.LastLogicalOffset
```
Comment (ramymedhat) [inline go/pkg/client/bytestream.go:90]: Doesn't look like you're using the initialOffset in the first write?

**Review by ramymedhat — COMMENTED** [review]
Can you clarify how you tested this? Also can you add unit tests?

File: go/pkg/client/bytestream.go (hunk 2 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go:90]: No it's no being used. I created initialOffset to update offset after writing in line 127.

**Comment by YukaSadaoka** [comment]
> Can you clarify how you tested this? Also can you add unit tests?
I used logstream integration and prober tests to test the changes. 
Acknowledged adding unit tests for this. I will add it and will update here once I'm done.

File: go/pkg/client/bytestream.go (hunk 2 above)
Comment (ramymedhat) [reply go/pkg/client/bytestream.go:90]: I'm confused. The chunker has chunked the entire blob, right? So its actually disregarding the initial offset? How will you make sure that you're uploading the right chunk relative to the desired offset?

File: go/pkg/client/client.go [hunk 3]
```diff
@@ -474,6 +474,12 @@ func getRPCCreds(ctx context.Context, credFile string, useApplicationDefault boo
 	return rpcCreds, CredsFileAuth, nil
 }
 
+// ByteStreamWriteOpts sets options required for resumable ByteStream write operations.
+type ByteStreamWriteOpts struct {
```
Comment (mrahs) [inline go/pkg/client/client.go:478]: Have you considered using [variadic options](https://google.github.io/styleguide/go/best-practices#variadic-options)? It allows you to reuse the existing method without affecting existing call sites.

File: go/pkg/client/client.go [hunk 4]
```diff
@@ -474,6 +474,12 @@ func getRPCCreds(ctx context.Context, credFile string, useApplicationDefault boo
...
+	FinishWrite       bool
```
Comment (mrahs) [inline go/pkg/client/client.go:479]: What is the use-case for this option?

It looks like the new method sends all the given bytes to the server and tells the server to start writing at a specific offset. Since all the given bytes are sent, `req.FinishWrite` should always be sent at the end. No?

Comment (YukaSadaoka) [reply go/pkg/client/client.go:479]: The boolean flag is for the case where clients didn't send all bytes in the previous write requests and want to resume writing the rest of bytes. Yes it's always sent but I wanted to allow clients to set the flag based on their remaining data so that they can resume writing.

File: go/pkg/client/client.go (hunk 3 above)
Comment (YukaSadaoka) [reply go/pkg/client/client.go:478]: Thanks for the input. I have considered to use it but adding my changes to the current existing method(WriteBytes in bytestream.go) seems making readability not great. Would you recommend me to use variadic options over an option struct?

File: go/pkg/client/bytestream.go (hunk 2 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go:90]: Right the current implementation is disregarding the initial offset(offset property in Chunker), I think I can update the code to use the offset property for a clarity. Let me try to update the code.

File: go/pkg/client/client.go (hunk 4 above)
Comment (mrahs) [reply go/pkg/client/client.go:479]: That makes sense. Just noticed that the client associates the flag with the given resource name. Your implementation allows users to call the public method multiple times for the same resource name without causing errors, which makes sense with the new arbitrary offset option.

File: go/pkg/client/client.go (hunk 3 above)
Comment (mrahs) [reply go/pkg/client/client.go:478]: The new method shares most of its code with the existing one, which seems like a good candidate for an option parameter. Variadic options allow backwards compatibility with existing call sites.

I'm thinking something like:
```go
type ByteStreamWriteOption interface {
  Apply(*ByteStreamWriteOpts)
}
type ByteStreamOptFinishWrite bool
func (o ByteSteramOptFinishWrite) Apply(opts *ByteStreamWriteOpts) {
  opts.FinishWrite = bool(o)
}

func (c *Client) WriteBytesWithOffset(ctx context.Context, name string, data []byte, rawOpts ...ByteStreamWriteOption) (int64, error) {
  opts := &ByteStreamWriteOpts{}
  for _, o := range rawOpts {
    o.Apply(opts)
  }
  ...
}
```

File: go/pkg/client/bytestream.go [hunk 5]
```diff
@@ -73,6 +73,61 @@ func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chun
...
+func (c *Client) WriteBytesWithOffset(ctx context.Context, name string, data []byte, opts *ByteStreamWriteOpts) (int64, error) {
+	ue := uploadinfo.EntryFromBlob(data)
+	ch, err := chunker.New(ue, false, int(c.ChunkMaxSize))
+	if err != nil {
+		return 0, err
+	}
+	return c.writeChunkedWithOffset(ctx, name, ch, opts)
+}
+
+// writeChunkedWithOffset uploads chunked data with a given resource name to the CAS at an arbitrary offset.
```
Comment (mrahs) [inline go/pkg/client/bytestream.go:86]: I think this comment is worth having for the exported method to expose it to IDEs and users.

File: go/pkg/client/client.go (hunk 3 above)
Comment (YukaSadaoka) [reply go/pkg/client/client.go:478]: Update the code per your suggestion. Please let me know if I'm missing something.

File: go/pkg/client/bytestream.go (hunk 5 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Updated the exported method's comment.

File: go/pkg/client/client.go (hunk 3 above)
Comment (mrahs) [reply go/pkg/client/client.go:478]: My bad, I meant to write the original public method name `WriteBytes` rather than add another version:
```go
func (c *Client) WriteBytes(ctx context.Context, name string, data []byte, rawOpts ...ByteStreamWriteOption) (int64, error) {
  ...
}

File: go/pkg/client/bytestream.go [hunk 6]
```diff
@@ -26,33 +48,56 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
...
+		return 0, err
+	}
+	return c.writeChunked(ctx, name, ch, opts...)
+}
+
 // writeChunked uploads chunked data with a given resource name to the CAS.
-func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker) (int64, error) {
+func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker, rawOpts ...ByteStreamWriteOption) (int64, error) {
+	if len(rawOpts) == 1 {
+		return 0, errors.Errorf("Missing an arugment")
```
Comment (mrahs) [inline go/pkg/client/bytestream.go:65]: Doesn't this break `WriteBytes` since it shares this internal method without the new options?

File: go/pkg/client/bytestream.go [hunk 7]
```diff
@@ -26,33 +48,56 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
...
 			if err != nil {
 				return err
 			}
-			if chunk.Offset == 0 {
-				req.ResourceName = name
-			}
 			req.WriteOffset = chunk.Offset
+			// If LastLogicaloffset is provided as an option, we want uploading bytes
+			// at the last logcial offset + the chunk offset which is relative to Chunker passed in.
+			if resume {
```
Comment (mrahs) [inline go/pkg/client/bytestream.go:95]: Is this condition necessary? The default value for `opts.LastLogicalOffset` is `0`, which means `chunk.Offset + opts.LastLogicalOffset` should give the correct offset in both cases.

File: go/pkg/client/bytestream.go [hunk 8]
```diff
@@ -26,33 +48,56 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
...
+	return c.writeChunked(ctx, name, ch, opts...)
+}
+
 // writeChunked uploads chunked data with a given resource name to the CAS.
-func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker) (int64, error) {
+func (c *Client) writeChunked(ctx context.Context, name string, ch *chunker.Chunker, rawOpts ...ByteStreamWriteOption) (int64, error) {
+	if len(rawOpts) == 1 {
+		return 0, errors.Errorf("Missing an arugment")
+	}
+	opts := &ByteStreamWriteOpts{}
```
Comment (mrahs) [inline go/pkg/client/bytestream.go:67]: The `FinishWrite` option should default to `true` to maintain current semantics. That is, we'll always finish at the end of the stream unless instructed otherwise.

File: go/pkg/client/bytestream.go [hunk 9]
```diff
@@ -26,33 +48,56 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
...
+	}
+
 	var totalBytes int64
 	closure := func() error {
 		// Retry by starting the stream from the beginning.
 		if err := ch.Reset(); err != nil {
 			return errors.Wrap(err, "failed to Reset")
 		}
 		totalBytes = int64(0)
-		// TODO(olaola): implement resumable uploads.
```
Comment (mrahs) [inline go/pkg/client/bytestream.go:38]: Hmm, this implementation is not fully resumable since retries will resend all the chunks. A true resumable implementation would need to track the offset across retries to avoid resending successfully sent chunks. If that is not in the scope of your change, you could simply clarify that in the doc comment (that the new option allows to upload to an arbitrary offset,  but retries still resend from the specified offset).

File: go/pkg/client/bytestream.go (hunk 6 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Resolved this with Anas offline. I removed this code block after refactoring.

File: go/pkg/client/bytestream.go (hunk 8 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Done.

File: go/pkg/client/bytestream.go (hunk 9 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go:38]: Done.

File: go/pkg/client/bytestream_test.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+	"testing"
+
+	"google.golang.org/grpc"
+
+	bsgrpc "google.golang.org/genproto/googleapis/bytestream"
+	bspb "google.golang.org/genproto/googleapis/bytestream"
+)
+
+var (
+	logStreams    map[string]*logStream
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:16]: This can be a member of the `ByteStream` strucrt on

File: go/pkg/client/bytestream_test.go [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+	logStreamData = []byte("Hello World! This is large data to send.")
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:16]: The value of this variable is constant throughout its lifetime. Might as well make it official by declaring as a constant.

File: go/pkg/client/bytestream_test.go [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+	logicalOffset int64
+	finalized     bool
+}
+
+func TestWriteBytesWithOffsetSuccess_LogStream(t *testing.T) {
+	tests := []struct {
+		description string
+		ls          *logStream
+		data        []byte
+		numIterate  int
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:30]: Nit: I think this could have a more descriptive name that makes it immediately obvious that 1) it is related to the field data and 2) it specifies how many parts that data will be split into. Maybe something like `dataPartsLen` or `dataPartsCount`?

File: go/pkg/client/bytestream_test.go [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+		opts        []ByteStreamWriteOption
+		wantBytes   int64
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:32]: Nit: `wantBytes` sounds like a `[]byte` value. `wantBytesLen` might be a less ambiguous name.

File: go/pkg/client/bytestream_test.go [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+		},
+	}
+
+	b := setup(t)
+	defer b.shutDown()
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			size := len(test.data)/test.numIterate + 1
+			var start int
+			start = int(test.opts[0].(ByteStreamOptOffset))
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:74]: Nit: Those two lines can be joined into one using a short declaration: `start := int(...)`.

File: go/pkg/client/bytestream_test.go [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+		description string
+		ls          *logStream
+		data        []byte
+		numIterate  int
+		opts        []ByteStreamWriteOption
+		wantBytes   int64
+	}{
+		{
+			description: "valid data with offset 0",
+			ls:          &logStream{logStreamID: "logstream1", logicalOffset: 0},
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:36]: Nit: `logStreamID` can probably just reuse the `test.description`. That way, it's less likely that a future developer might forget to change the id to something unique after copying an existing test case. The test description is likely to be unique in the same test set (see my comment about encapsulating the streams map).

File: go/pkg/client/bytestream_test.go [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+				writtenBytes, err := b.client.WriteBytesWithOptions(b.ctx, lsID, test.data[start:end], test.opts...)
+				if err != nil {
+					t.Errorf("WriteBytesWithOffset() failed unexpectedly: %v", err)
+				}
+				test.opts[0] = test.opts[0].(ByteStreamOptOffset) + ByteStreamOptOffset(writtenBytes)
+				start = end
+				end += size
+			}
+
+			if logStreams[lsID].logicalOffset != test.wantBytes {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go]: It's worth asserting a similar condition within the parts loop above. That is, after each call the stream's offset is incremented by the part's length.

File: go/pkg/client/bytestream_test.go [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+				t.Errorf("WriteBytesWithOffset() = %d, want %d", logStreams[lsID].logicalOffset, test.wantBytes)
+			}
+			if logStreams[lsID].finalized == false {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go]: Same here as well. It's worth asserting this condition within the parts loop.

File: go/pkg/client/bytestream_test.go (hunk 10 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go [hunk 18]
```diff
@@ -0,0 +1,294 @@
...
+		})
+	}
+}
+
+func TestWriteBytesWithOffsetErrors_LogStream(t *testing.T) {
+	tests := []struct {
+		description string
+		ls          *logStream
+		data        []byte
+		notFound    bool
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:112]: I couldn't figure out how this field is used. If you're trying to test for a non-existing stream, I might suggest making the `ls` field of the corresponding test `nil`.

File: go/pkg/client/bytestream_test.go [hunk 19]
```diff
@@ -0,0 +1,294 @@
...
+			ChunkMaxSize(dataSize).Apply(b.client)
+
+			_, err := b.client.WriteBytesWithOptions(b.ctx, lsID, logStreamData[int(offset.(ByteStreamOptOffset)):], offset)
+			if err != nil {
+				t.Errorf("WriteBytesWithOffset() failed unexpectedly: %v", err)
+			}
+			if logStreams[lsID].logicalOffset != int64(dataSize) {
+				t.Errorf("WriteBytesWithOffset() = %d, want %d", logStreams[lsID].logicalOffset, int64(dataSize))
+			}
+			if logStreams[lsID].finalized != true {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go]: Nit: `!= true` is redundant.

File: go/pkg/client/bytestream_test.go [hunk 20]
```diff
@@ -0,0 +1,294 @@
...
+
+type Server struct {
+	client   *Client
+	listener net.Listener
+	server   *grpc.Server
+	fake     *Bytestream
+	ctx      context.Context
+}
+
+func setup(t *testing.T) *Server {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:235]: `setup` is a generic name. The implementation seems to be doing `newServer` or `newFakeServer`.

File: go/pkg/client/retries_test.go [hunk 21]
```diff
@@ -285,6 +285,36 @@ func TestWriteRetries(t *testing.T) {
 	}
 }
 
+func TestWriteRetriesWithOptions(t *testing.T) {
+	opts := [][]client.ByteStreamWriteOption{
+		{client.ByteSteramOptFinishWrite(true), client.ByteStreamOptOffset(0)},
+		{client.ByteSteramOptFinishWrite(false), client.ByteStreamOptOffset(0)},
+		{client.ByteSteramOptFinishWrite(true)},
+		{client.ByteSteramOptFinishWrite(false)},
+		{client.ByteStreamOptOffset(0)},
```
Comment (mrahs) [inline go/pkg/client/retries_test.go:294]: Can you add a test case with a non-zero offset?

File: go/pkg/client/bytestream_test.go (hunk 12 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 13 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 14 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 15 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done. Updated the logStreamID in TestWriteBytesWithOffsetSuccess_LogStream and TestWriteBytesWithOffsetErrors_LogStream.

File: go/pkg/client/bytestream_test.go (hunk 17 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 18 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 19 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 20 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/retries_test.go (hunk 21 above)
Comment (YukaSadaoka) [reply go/pkg/client/retries_test.go]: Done.

File: go/pkg/client/bytestream_test.go [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+
+				writtenBytes, err := b.client.WriteBytesWithOptions(b.ctx, lsID, test.data[start:end], test.opts...)
+				if err != nil {
+					t.Errorf("WriteBytesWithOffset() failed unexpectedly: %v", err)
+				}
+				if b.fake.logStreams[lsID].logicalOffset != int64(end) {
+					t.Errorf("WriteBytesWithOffset() = %d, want %d", b.fake.logStreams[lsID].logicalOffset, end)
+				}
+				// LogStream shouldn't be finalized when we set ByteSteramOptFinishWrite false.
+				if i != test.dataPartsLen-1 && b.fake.logStreams[lsID].finalized == true {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:94]:
```suggestion
				if i != test.dataPartsLen-1 && b.fake.logStreams[lsID].finalized {
```

File: go/pkg/client/bytestream_test.go [hunk 23]
```diff
@@ -0,0 +1,295 @@
...
+
+				test.opts[0] = test.opts[0].(ByteStreamOptOffset) + ByteStreamOptOffset(writtenBytes)
+				start = end
+				end += size
+			}
+
+			if b.fake.logStreams[lsID].logicalOffset != test.wantBytesLen {
+				t.Errorf("WriteBytesWithOffset() = %d, want %d", b.fake.logStreams[lsID].logicalOffset, test.wantBytesLen)
+			}
+			if b.fake.logStreams[lsID].finalized == false {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:106]:
```suggestion
			if !b.fake.logStreams[lsID].finalized {
```

File: go/pkg/client/bytestream_test.go (hunk 11 above)
Comment (mrahs) [reply go/pkg/client/bytestream_test.go]: Right. I must have thought of it as a string constant. Not a big deal.

File: go/pkg/client/bytestream.go [hunk 24]
```diff
@@ -26,33 +48,52 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
...
+			req := &bspb.WriteRequest{ResourceName: name}
 			chunk, err := ch.Next()
 			if err != nil {
 				return err
 			}
-			if chunk.Offset == 0 {
-				req.ResourceName = name
-			}
-			req.WriteOffset = chunk.Offset
+			req.WriteOffset = chunk.Offset + opts.LastLogicalOffset
```
Comment (ramymedhat) [inline go/pkg/client/bytestream.go:93]: This would apply to all chunks?

File: go/pkg/client/bytestream_test.go (hunk 22 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 23 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream.go [hunk 25]
```diff
@@ -15,6 +15,28 @@ import (
...
 )
 
+// ByteStreamWriteOption sets options required for resumable upload.
+type ByteStreamWriteOption interface {
+	Apply(*ByteStreamWriteOpts)
+}
+
+// ByteSteramOptFinishWrite is a boolean flag to set FinishWrite
+// in ByteStream.WriteRequest.
+type ByteSteramOptFinishWrite bool
```
[the remark this answers is not in the document]
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Done.

File: go/pkg/client/bytestream.go (hunk 24 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go:93]: Yes but LastLogicalOffset doesn't basically affect WriteOffset in the case where LastLogicalOffset is not passed in as ByteStreamWriteOption. If LastLogicalOffset is not passed in, the value of LastLogicalOffset will be 0 as default and it will be basically req.WriteOffset = chunk.Offset. 

Refactored this code block in this way so that it's easier to read and avoid adding many if statements.

File: go/pkg/client/bytestream_test.go [hunk 26]
```diff
@@ -0,0 +1,295 @@
...
+
+	}
+}
+
+func TestWirteBytesWithOptions_LogStream_Offset(t *testing.T) {
+	b := newServer(t)
+	defer b.shutDown()
+	dataSize := len(logStreamData)
+	for _, offset := range []ByteStreamWriteOption{ByteStreamOptOffset(0), ByteStreamOptOffset(6)} {
+		t.Run(fmt.Sprintf("ByteSteramOptFinishWrite=%t", offset), func(t *testing.T) {
```
Comment (mrahs) [inline go/pkg/client/bytestream_test.go:205]: Typo
```suggestion
		t.Run(fmt.Sprintf("ByteStreamOptFinishWrite=%t", offset), func(t *testing.T) {
```

File: go/pkg/client/bytestream.go (hunk 24 above)
Comment (ramymedhat) [reply go/pkg/client/bytestream.go:93]: Ack

File: go/pkg/client/bytestream.go [hunk 27]
```diff
@@ -26,33 +48,52 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
 	return err
 }
 
+// WriteBytesWithOptions uploads a byte slice with a given resource name to the CAS
+// at an arbitrary offset but retries still resend from the specified offset.
+func (c *Client) WriteBytesWithOptions(ctx context.Context, name string, data []byte, opts ...ByteStreamWriteOption) (int64, error) {
```
Comment (ramymedhat) [inline go/pkg/client/bytestream.go:53]: Can you just pass in the offset and finishWrite here? I think that would be simpler since its only two args.

File: go/pkg/client/bytestream_test.go [hunk 28]
```diff
@@ -0,0 +1,295 @@
...
+			wantBytesLen: 0,
+		},
+	}
+
+	b := newServer(t)
+	defer b.shutDown()
+	for _, test := range tests {
+		t.Run(test.description, func(t *testing.T) {
+			size := len(test.data)/test.dataPartsLen + 1
+			start := int(test.opts[0].(ByteStreamOptOffset))
```
Comment (ramymedhat) [inline go/pkg/client/bytestream_test.go:71]: Just using function parameters rather than an opt struct will make this simpler as well.

File: go/pkg/client/bytestream_test.go [hunk 29]
```diff
@@ -0,0 +1,295 @@
...
+				t.Error("WriteBytesWithOffset() got nil error, want non-nil error")
+			}
+			if writtenBytes != 0 {
+				t.Errorf("WriteBytesWithOffset() got %d byte(s), want 0 byte", writtenBytes)
+			}
+		})
+	}
+}
+
+func TestWirteBytesWithOptions_LogStream_FinishWrite(t *testing.T) {
```
Comment (ramymedhat) [inline go/pkg/client/bytestream_test.go:174]: Write

File: go/pkg/client/bytestream_test.go [hunk 30]
```diff
@@ -0,0 +1,295 @@
...
+			opts:        []ByteStreamWriteOption{ByteStreamOptOffset(0), ByteStreamOptFinishWrite(false)},
+		},
+		{
+			description: "not found",
+			ls:          nil,
+			data:        logStreamData,
+			opts:        []ByteStreamWriteOption{ByteStreamOptOffset(0), ByteStreamOptFinishWrite(false)},
+		},
+		{
+			description: "invlid smaller offset",
```
Comment (ramymedhat) [inline go/pkg/client/bytestream_test.go:133]: s/invlid/invalid here and below

File: go/pkg/client/bytestream.go [hunk 31]
```diff
@@ -26,33 +48,52 @@ func (c *Client) WriteBytes(ctx context.Context, name string, data []byte) error
 	return err
 }
 
+// WriteBytesWithOptions uploads a byte slice with a given resource name to the CAS
+// at an arbitrary offset but retries still resend from the specified offset.
```
Comment (ramymedhat) [inline go/pkg/client/bytestream.go:52]: s/offset/remote offset?

File: go/pkg/client/bytestream.go (hunk 27 above)
Comment (ramymedhat) [inline go/pkg/client/bytestream.go:53]: s/WriteBytesWithOptions/WriteBytesAtRemoteOffset? I think that makes it clearer what this is doing.

Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go:53]: In case if we add more options to the method(or to writeChunked method) in the future and this method keeps a variadic option, we can add them easily without affecting the call sites of this method. I'm not sure how likely we will add more options but I think that scales better. What do you think?

File: go/pkg/client/bytestream_test.go (hunk 26 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Removed the unit test method as the implementation has changed a bit.

File: go/pkg/client/bytestream.go (hunk 27 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Resolved offline. Updated the method signatures.

File: go/pkg/client/bytestream_test.go (hunk 28 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 29 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream_test.go (hunk 30 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream_test.go]: Done.

File: go/pkg/client/bytestream.go (hunk 27 above)
Comment (YukaSadaoka) [reply go/pkg/client/bytestream.go]: Done.

**Review by ramymedhat — APPROVED** [review]
